examples/sui-move/escrow/sources/shared_example.move

```diff
@@ -0,0 +1,290 @@
+
+        transfer::public_share_object(
+            Exchange<T> {
+                id, sender, recipient, exchange_for, escrowed
+            }
+        );
+    }
+
+    /// The `recipient` of the escrow can exchange `obj` with the escrowed item
+    public fun exchange<T: key + store, ExchangeForT: key + store>(
+        obj: ExchangeForT,
+        escrow: &mut Exchange<T>,
+        ctx: &TxContext
+    ) {
+        assert!(option::is_some(&escrow.escrowed), EAlreadyExchangedOrCancelled);
+        let escrowed_item = option::extract<T>(&mut escrow.escrowed);
+        assert!(&tx_context::sender(ctx) == &escrow.recipient, EMismatchedSenderRecipient);
+        assert!(object::borrow_id(&obj) == &escrow.exchange_for, EMismatchedExchangeObject);
+        // everything matches. do the swap!
+        transfer::public_transfer(escrowed_item, tx_context::sender(ctx));
+        transfer::public_transfer(obj, escrow.sender);
+    }
+
+    /// The `creator` can cancel the escrow and get back the escrowed item
+    public fun cancel<T: key + store>(
+        escrow: &mut Exchange<T>,
+        ctx: &TxContext
+    ): T {
+        assert!(&tx_context::sender(ctx) == &escrow.sender, EMismatchedSenderRecipient);
+        assert!(option::is_some(&escrow.escrowed), EAlreadyExchangedOrCancelled);
+        option::extract<T>(&mut escrow.escrowed)
+    }
+
+    // === Tests ===
+    use sui::coin::{Self, Coin};
+    use sui::sui::SUI;
+    #[test_only] use sui::test_scenario::{Self, Scenario};
+
+    #[test_only]
+    fun test_coin(ts: &mut Scenario): Coin<SUI> {
+        coin::mint_for_testing<SUI>(42, test_scenario::ctx(ts))
+    }
+
+    #[test]
+    fun test_successful_swap() {
+        let ts = test_scenario::begin(@0x0);
+        let alice = @0xA;
+        let bob = @0xB;
+
+        // Initialize coins for Alice + Bob
+        let (alice_coin_id) = {
+            test_scenario::next_tx(&mut ts, alice);
+            let c = test_coin(&mut ts);
+            let alice_coin_id = (object::id(&c));
+            transfer::public_transfer(c, alice);
+            alice_coin_id
+        };
+
+        let (bob_coin_id) = {
+            test_scenario::next_tx(&mut ts, bob);
+            let c = test_coin(&mut ts);
+            let bob_coin_id = (object::id(&c));
+            transfer::public_transfer(c, bob);
+            bob_coin_id
+        };
+        // Alice creates a public Exchange asking for an object from bob
+        {
+            test_scenario::next_tx(&mut ts, alice);
+            let a_coin: Coin<SUI> = test_scenario::take_from_sender(&ts);
+            create<Coin<SUI>>(
+                bob,
+                bob_coin_id,
+                a_coin,
+                test_scenario::ctx(&mut ts)
+            );
+        };
+
+        // Bob pulls the shared object and exchanges and transfers to himself
+        {
+            test_scenario::next_tx(&mut ts, bob);
+            let exchange_obj = test_scenario::take_shared<Exchange<Coin<SUI>>>(&mut ts);
```

Comment:
nit: Use the pattern you've used on the line below to annotate the variable instead of provide the type parameter?

```suggestion
            let exchange_obj: Exchange<Coin<SUI>> = test_scenario::take_shared(&mut ts);
```

In Move, this is mostly a personal stylistic preference thing, because these two patterns are as noisy as each other, but in Rust, annotating the variable is slightly less noisy.  I tend to prefer annotating the variable for clarity and for consistency with recommendations in Rust, but this is not a big deal.  If you prefer the other way though, better to be consistent throughout (i.e. don't instantiate the type parameter on one line, and then annotate the variable on the next line).